File: ClearScript/V8/ClearScriptV8/V8ObjectImpl.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

#include "ClearScriptV8Managed.h"

namespace Microsoft {
namespace ClearScript {
namespace V8 {

    //-------------------------------------------------------------------------
    // local helper functions
    //-------------------------------------------------------------------------

    static void ArrayBufferOrViewDataCallback(void* pvData, void* pvArg)
    {
        (*static_cast<Action<IntPtr>^*>(pvArg))(IntPtr(pvData));
    }

    //-------------------------------------------------------------------------
    // V8ObjectImpl implementation
    //-------------------------------------------------------------------------

    V8ObjectImpl::V8ObjectImpl(V8ObjectHolder* pHolder, V8Value::Subtype subtype):
        m_gcLock(gcnew Object),
        m_pspHolder(new SharedPtr<V8ObjectHolder>(pHolder)),
        m_Subtype(subtype)
    {
    }

    //-------------------------------------------------------------------------

    Object^ V8ObjectImpl::GetProperty(String^ gcName)
    {
        try
        {
            return V8ContextProxyImpl::ExportValue(V8ObjectHelpers::GetProperty(GetHolder(), StdString(gcName)));
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    void V8ObjectImpl::SetProperty(String^ gcName, Object^ gcValue)
    {
        try
        {
            V8ObjectHelpers::SetProperty(GetHolder(), StdString(gcName), V8ContextProxyImpl::ImportValue(gcValue));
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    bool V8ObjectImpl::DeleteProperty(String^ gcName)
    {
        try
        {
            return V8ObjectHelpers::DeleteProperty(GetHolder(), StdString(gcName));
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    array<String^>^ V8ObjectImpl::GetPropertyNames()
    {
        try
        {
            std::vector<StdString> names;
            V8ObjectHelpers::GetPropertyNames(GetHolder(), names);
            auto nameCount = static_cast<int>(names.size());

            auto gcNames = gcnew array<String^>(nameCount);
            for (auto index = 0; index < nameCount; index++)
            {
                gcNames[index] = names[index].ToManagedString();
            }

            return gcNames;
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    Object^ V8ObjectImpl::GetProperty(int index)
    {
        try
        {
            return V8ContextProxyImpl::ExportValue(V8ObjectHelpers::GetProperty(GetHolder(), index));
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    void V8ObjectImpl::SetProperty(int index, Object^ gcValue)
    {
        try
        {
            V8ObjectHelpers::SetProperty(GetHolder(), index, V8ContextProxyImpl::ImportValue(gcValue));
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    bool V8ObjectImpl::DeleteProperty(int index)
    {
        try
        {
            return V8ObjectHelpers::DeleteProperty(GetHolder(), index);
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    array<int>^ V8ObjectImpl::GetPropertyIndices()
    {
        try
        {
            std::vector<int> indices;
            V8ObjectHelpers::GetPropertyIndices(GetHolder(), indices);
            auto indexCount = static_cast<int>(indices.size());

            auto gcIndices = gcnew array<int>(indexCount);
            for (auto index = 0; index < indexCount; index++)
            {
                gcIndices[index] = indices[index];
            }

            return gcIndices;
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    Object^ V8ObjectImpl::Invoke(bool asConstructor, array<Object^>^ gcArgs)
    {
        try
        {
            std::vector<V8Value> importedArgs;
            ImportValues(gcArgs, importedArgs);

            return V8ContextProxyImpl::ExportValue(V8ObjectHelpers::Invoke(GetHolder(), asConstructor, importedArgs));
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    Object^ V8ObjectImpl::InvokeMethod(String^ gcName, array<Object^>^ gcArgs)
    {
        try
        {
            std::vector<V8Value> importedArgs;
            ImportValues(gcArgs, importedArgs);

            return V8ContextProxyImpl::ExportValue(V8ObjectHelpers::InvokeMethod(GetHolder(), StdString(gcName), importedArgs));
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    bool V8ObjectImpl::IsArrayBufferOrView()
    {
        return m_Subtype != V8Value::Subtype::None;
    }

    //-------------------------------------------------------------------------

    V8ArrayBufferOrViewKind V8ObjectImpl::GetArrayBufferOrViewKind()
    {
        auto kind = V8ArrayBufferOrViewKind::None;

        if (m_Subtype == V8Value::Subtype::ArrayBuffer)
            kind = V8ArrayBufferOrViewKind::ArrayBuffer;
        else if (m_Subtype == V8Value::Subtype::DataView)
            kind = V8ArrayBufferOrViewKind::DataView;
        else if (m_Subtype == V8Value::Subtype::Uint8Array)
            kind = V8ArrayBufferOrViewKind::Uint8Array;
        else if (m_Subtype == V8Value::Subtype::Uint8ClampedArray)
            kind = V8ArrayBufferOrViewKind::Uint8ClampedArray;
        else if (m_Subtype == V8Value::Subtype::Int8Array)
            kind = V8ArrayBufferOrViewKind::Int8Array;
        else if (m_Subtype == V8Value::Subtype::Uint16Array)
            kind = V8ArrayBufferOrViewKind::Uint16Array;
        else if (m_Subtype == V8Value::Subtype::Int16Array)
            kind = V8ArrayBufferOrViewKind::Int16Array;
        else if (m_Subtype == V8Value::Subtype::Uint32Array)
            kind = V8ArrayBufferOrViewKind::Uint32Array;
        else if (m_Subtype == V8Value::Subtype::Int32Array)
            kind = V8ArrayBufferOrViewKind::Int32Array;
        else if (m_Subtype == V8Value::Subtype::Float32Array)
            kind = V8ArrayBufferOrViewKind::Float32Array;
        else if (m_Subtype == V8Value::Subtype::Float64Array)
            kind = V8ArrayBufferOrViewKind::Float64Array;

        return kind;
    }
    
    //-------------------------------------------------------------------------

    V8ArrayBufferOrViewInfo^ V8ObjectImpl::GetArrayBufferOrViewInfo()
    {
        try
        {
            auto kind = GetArrayBufferOrViewKind();
            if (kind != V8ArrayBufferOrViewKind::None)
            {
                V8Value arrayBuffer(V8Value::Null);
                size_t offset;
                size_t size;
                size_t length;

                V8ObjectHelpers::GetArrayBufferOrViewInfo(GetHolder(), arrayBuffer, offset, size, length);
                return gcnew V8ArrayBufferOrViewInfo(kind, (IV8Object^)V8ContextProxyImpl::ExportValue(arrayBuffer), offset, size, length);
            }

            return nullptr;
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    void V8ObjectImpl::InvokeWithArrayBufferOrViewData(Action<IntPtr>^ gcAction)
    {
        try
        {
            V8ObjectHelpers::InvokeWithArrayBufferOrViewData(GetHolder(), ArrayBufferOrViewDataCallback, &gcAction);
        }
        catch (const V8Exception& exception)
        {
            exception.ThrowScriptEngineException();
        }
    }

    //-------------------------------------------------------------------------

    SharedPtr<V8ObjectHolder> V8ObjectImpl::GetHolder()
    {
        BEGIN_LOCK_SCOPE(m_gcLock)

            if (m_pspHolder == nullptr)
            {
                throw gcnew ObjectDisposedException(ToString());
            }

            return *m_pspHolder;

        END_LOCK_SCOPE
    }

    //-------------------------------------------------------------------------

    V8Value::Subtype V8ObjectImpl::GetSubtype()
    {
        return m_Subtype;
    }

    //-------------------------------------------------------------------------

    V8ObjectImpl::~V8ObjectImpl()
    {
        SharedPtr<V8ObjectHolder> spHolder;

        BEGIN_LOCK_SCOPE(m_gcLock)

            if (m_pspHolder != nullptr)
            {
                // hold V8 object holder for destruction outside lock scope
                spHolder = *m_pspHolder;
                delete m_pspHolder;
                m_pspHolder = nullptr;
            }

        END_LOCK_SCOPE

        if (!spHolder.IsEmpty())
        {
            GC::SuppressFinalize(this);
        }
    }

    //-------------------------------------------------------------------------

    V8ObjectImpl::!V8ObjectImpl()
    {
        if (m_pspHolder != nullptr)
        {
            delete m_pspHolder;
            m_pspHolder = nullptr;
        }
    }

    //-------------------------------------------------------------------------

    void V8ObjectImpl::ImportValues(array<Object^>^ gcValues, std::vector<V8Value>& importedValues)
    {
        importedValues.clear();
        if (gcValues != nullptr)
        {
            auto valueCount = gcValues->Length;
            importedValues.reserve(valueCount);

            for (auto index = 0; index < valueCount; index++)
            {
                importedValues.push_back(V8ContextProxyImpl::ImportValue(gcValues[index]));
            }
        }
    }

}}}

File: ClearScript/V8/SplitProxy/V8ObjectImpl.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

using System;
using System.Runtime.CompilerServices;
using Microsoft.ClearScript.JavaScript;
using Microsoft.ClearScript.Util;

namespace Microsoft.ClearScript.V8.SplitProxy
{
    internal sealed class V8ObjectImpl : IV8Object
    {
        private V8EntityHolder holder;

        public V8Object.Handle Handle => (V8Object.Handle)holder.Handle;

        public V8ObjectImpl(V8Object.Handle hObject, V8Value.Subtype subtype, V8Value.Flags flags, int identityHash)
        {
            holder = new V8EntityHolder("V8 object", () => hObject);
            Subtype = subtype;
            Flags = flags;
            IdentityHash = identityHash;
        }

        public V8Value.Subtype Subtype { get; }

        public V8Value.Flags Flags { get; }

        #region IV8Object implementation

        public JavaScriptObjectKind ObjectKind
        {
            get
            {
                switch (Subtype)
                {
                    case V8Value.Subtype.Function:
                        return JavaScriptObjectKind.Function;

                    case V8Value.Subtype.Iterator:
                        return JavaScriptObjectKind.Iterator;

                    case V8Value.Subtype.Promise:
                        return JavaScriptObjectKind.Promise;

                    case V8Value.Subtype.Array:
                        return JavaScriptObjectKind.Array;

                    case V8Value.Subtype.ArrayBuffer:
                        return JavaScriptObjectKind.ArrayBuffer;

                    case V8Value.Subtype.DataView:
                        return JavaScriptObjectKind.DataView;

                    case V8Value.Subtype.Uint8Array:
                    case V8Value.Subtype.Uint8ClampedArray:
                    case V8Value.Subtype.Int8Array:
                    case V8Value.Subtype.Uint16Array:
                    case V8Value.Subtype.Int16Array:
                    case V8Value.Subtype.Uint32Array:
                    case V8Value.Subtype.Int32Array:
                    case V8Value.Subtype.BigUint64Array:
                    case V8Value.Subtype.BigInt64Array:
                    case V8Value.Subtype.Float32Array:
                    case V8Value.Subtype.Float64Array:
                        return JavaScriptObjectKind.TypedArray;

                    default:
                        return JavaScriptObjectKind.Unknown;
                }
            }
        }

        public JavaScriptObjectFlags ObjectFlags
        {
            get
            {
                var flags = JavaScriptObjectFlags.None;

                if (Flags.HasAllFlags(V8Value.Flags.Shared))
                {
                    flags |= JavaScriptObjectFlags.Shared;
                }

                if (Flags.HasAllFlags(V8Value.Flags.Async))
                {
                    flags |= JavaScriptObjectFlags.Async;
                }

                if (Flags.HasAllFlags(V8Value.Flags.Generator))
                {
                    flags |= JavaScriptObjectFlags.Generator;
                }

                if (Flags.HasAllFlags(V8Value.Flags.Pending))
                {
                    flags |= JavaScriptObjectFlags.Pending;
                }

                if (Flags.HasAllFlags(V8Value.Flags.Rejected))
                {
                    flags |= JavaScriptObjectFlags.Rejected;
                }

                return flags;
            }
        }

        public int IdentityHash { get; }

        public object GetProperty(string name)
        {
            return V8SplitProxyNative.Invoke(static (instance, ctx) => instance.V8Object_GetNamedProperty(ctx.Handle, ctx.name), (Handle, name));
        }

        public bool TryGetProperty(string name, out object value)
        {
            var ctx = (Handle, name, value: (object)null);

            var result = V8SplitProxyNative.Invoke(
                static (instance, pCtx) =>
                {
                    ref var ctx = ref pCtx.AsRef();
                    return instance.V8Object_TryGetNamedProperty(ctx.Handle, ctx.name, out ctx.value);
                },
                StructPtr.FromRef(ref ctx)
            );

            value = ctx.value;
            return result;
        }

        public void SetProperty(string name, object value)
        {
            V8SplitProxyNative.Invoke(static (instance, ctx) => instance.V8Object_SetNamedProperty(ctx.Handle, ctx.name, ctx.value), (Handle, name, value));
        }

        public bool DeleteProperty(string name)
        {
            return V8SplitProxyNative.Invoke(static (instance, ctx) => instance.V8Object_DeleteNamedProperty(ctx.Handle, ctx.name), (Handle, name));
        }

        public string[] GetPropertyNames(bool includeIndices)
        {
            return V8SplitProxyNative.Invoke(static (instance, ctx) => instance.V8Object_GetPropertyNames(ctx.Handle, ctx.includeIndices), (Handle, includeIndices));
        }

        public object GetProperty(int index)
        {
            return V8SplitProxyNative.Invoke(static (instance, ctx) => instance.V8Object_GetIndexedProperty(ctx.Handle, ctx.index), (Handle, index));
        }

        public void SetProperty(int index, object value)
        {
            V8SplitProxyNative.Invoke(static (instance, ctx) => instance.V8Object_SetIndexedProperty(ctx.Handle, ctx.index, ctx.value), (Handle, index, value));
        }

        public bool DeleteProperty(int index)
        {
            return V8SplitProxyNative.Invoke(static (instance, ctx) => instance.V8Object_DeleteIndexedProperty(ctx.Handle, ctx.index), (Handle, index));
        }

        public int[] GetPropertyIndices()
        {
            return V8SplitProxyNative.Invoke(static (instance, handle) => instance.V8Object_GetPropertyIndices(handle), Handle);
        }

        public object Invoke(bool asConstructor, object[] args)
        {
            return V8SplitProxyNative.Invoke(static (instance, ctx) => instance.V8Object_Invoke(ctx.Handle, ctx.asConstructor, ctx.args), (Handle, asConstructor, args));
        }

        public object InvokeMethod(string name, object[] args)
        {
            return V8SplitProxyNative.Invoke(static (instance, ctx) => instance.V8Object_InvokeMethod(ctx.Handle, ctx.name, ctx.args), (Handle, name, args));
        }

        public bool IsPromise => Subtype == V8Value.Subtype.Promise;

        public bool IsArray => Subtype == V8Value.Subtype.Array;

        public bool IsShared => Flags.HasAllFlags(V8Value.Flags.Shared);

        public bool IsArrayBufferOrView
        {
            get
            {
                switch (Subtype)
                {
                    case V8Value.Subtype.ArrayBuffer:
                    case V8Value.Subtype.DataView:
                    case V8Value.Subtype.Uint8Array:
                    case V8Value.Subtype.Uint8ClampedArray:
                    case V8Value.Subtype.Int8Array:
                    case V8Value.Subtype.Uint16Array:
                    case V8Value.Subtype.Int16Array:
                    case V8Value.Subtype.Uint32Array:
                    case V8Value.Subtype.Int32Array:
                    case V8Value.Subtype.BigUint64Array:
                    case V8Value.Subtype.BigInt64Array:
                    case V8Value.Subtype.Float32Array:
                    case V8Value.Subtype.Float64Array:
                        return true;

                    default:
                        return false;
                }
            }
        }

        public V8ArrayBufferOrViewKind ArrayBufferOrViewKind
        {
            get
            {
                switch (Subtype)
                {
                    case V8Value.Subtype.ArrayBuffer:
                        return V8ArrayBufferOrViewKind.ArrayBuffer;

                    case V8Value.Subtype.DataView:
                        return V8ArrayBufferOrViewKind.DataView;

                    case V8Value.Subtype.Uint8Array:
                        return V8ArrayBufferOrViewKind.Uint8Array;

                    case V8Value.Subtype.Uint8ClampedArray:
                        return V8ArrayBufferOrViewKind.Uint8ClampedArray;

                    case V8Value.Subtype.Int8Array:
                        return V8ArrayBufferOrViewKind.Int8Array;

                    case V8Value.Subtype.Uint16Array:
                        return V8ArrayBufferOrViewKind.Uint16Array;

                    case V8Value.Subtype.Int16Array:
                        return V8ArrayBufferOrViewKind.Int16Array;

                    case V8Value.Subtype.Uint32Array:
                        return V8ArrayBufferOrViewKind.Uint32Array;

                    case V8Value.Subtype.Int32Array:
                        return V8ArrayBufferOrViewKind.Int32Array;

                    case V8Value.Subtype.BigUint64Array:
                        return V8ArrayBufferOrViewKind.BigUint64Array;

                    case V8Value.Subtype.BigInt64Array:
                        return V8ArrayBufferOrViewKind.BigInt64Array;

                    case V8Value.Subtype.Float32Array:
                        return V8ArrayBufferOrViewKind.Float32Array;

                    case V8Value.Subtype.Float64Array:
                        return V8ArrayBufferOrViewKind.Float64Array;

                    default:
                        return V8ArrayBufferOrViewKind.None;
                }
            }
        }

        public V8ArrayBufferOrViewInfo GetArrayBufferOrViewInfo()
        {
            var kind = ArrayBufferOrViewKind;
            if (kind != V8ArrayBufferOrViewKind.None)
            {
                var ctx = (Handle, arrayBuffer: (IV8Object)null, offset: 0UL, size: 0UL, length: 0UL);

                V8SplitProxyNative.Invoke(
                    static (instance, pCtx) =>
                    {
                        ref var ctx = ref pCtx.AsRef();
                        instance.V8Object_GetArrayBufferOrViewInfo(ctx.Handle, out ctx.arrayBuffer, out ctx.offset, out ctx.size, out ctx.length);
                    },
                    StructPtr.FromRef(ref ctx)
                );

                return new V8ArrayBufferOrViewInfo(kind, ctx.arrayBuffer, ctx.offset, ctx.size, ctx.length);
            }

            return null;
        }

        public void InvokeWithArrayBufferOrViewData(Action<IntPtr> action)
        {
            using (var actionScope = V8ProxyHelpers.CreateAddRefHostObjectScope(action))
            {
                var pAction = actionScope.Value;
                V8SplitProxyNative.Invoke(static (instance, ctx) => instance.V8Object_InvokeWithArrayBufferOrViewData(ctx.Handle, ctx.pAction), (Handle, pAction));
            }
        }

        public unsafe void InvokeWithArrayBufferOrViewData<TArg>(Action<IntPtr, TArg> action, in TArg arg)
        {
            var ctx = (action, arg);
            var pCtx = (IntPtr)Unsafe.AsPointer(ref ctx);

            Action<IntPtr, IntPtr> actionWithArg = static (pData, pCtx) =>
            {
                ref var ctx = ref Unsafe.AsRef<(Action<IntPtr, TArg> action, TArg arg)>(pCtx.ToPointer());
                ctx.action(pData, ctx.arg);
            };

            using (var actionScope = V8ProxyHelpers.CreateAddRefHostObjectScope(actionWithArg))
            {
                var pAction = actionScope.Value;
                V8SplitProxyNative.Invoke(static (instance, ctx) => instance.V8Object_InvokeWithArrayBufferOrViewDataWithArg(ctx.Handle, ctx.pAction, ctx.pCtx), (Handle, pAction, pCtx));
            }
        }

        #endregion

        #region disposal / finalization

        public void Dispose()
        {
            holder.ReleaseEntity();
            GC.KeepAlive(this);
        }

        ~V8ObjectImpl()
        {
            V8EntityHolder.Destroy(ref holder);
        }

        #endregion
    }
}
